iso-authoryear.bbx

```diff
@@ -2,23 +2,49 @@
 
 \RequireBibliographyStyle{iso}
 \ExecuteBibliographyOptions{%
-  sorting=nyt
+  sorting=nyt,
 }
 
-\def\printextrayear{%
-  \ifentrytype{patent}
-    {}
-    {\printfield{year}\printfield{extradate}}%
+% The field extradate turns from integer type to a letter
+% after year to distinguish more works of the author
+% from the same year, e.g. 1994a, 1994b.
+% If there is no date field in the bib entry,
+% we should state there is no date by printing
+% a localisation string, e.g. [n.d.].
+% We use 'labeldate' as it is also used in standard styles.
+\newbibmacro*{date+extradate}{%
+  \iffieldundef{year}
+    {\bibstring[\mkbibbrackets]{nodate}}
+    {\printlabeldateextra}%
```

Comment:
If we use `\printlabeldateextra` we shouldn't test for `year` here. We should test for `labelyear`. See
```latex
\documentclass{article}
\usepackage[style=iso-authoryear,backend=biber]{biblatex}

\begin{filecontents}[force]{\jobname.bib}
@book{appleby,
  author  = {Humphrey Appleby},
  title   = {On the Importance of the Civil Service},
  urldate = {2019-12-12},
}
\end{filecontents}
\addbibresource{\jobname.bib}

\begin{document}
\cite{appleby}
\printbibliography
\end{document}
```
**But** as I said before, I think it is pointless to test for a *nodate* situation, since `labeldate` can already handle that.
Conceptually I'd much prefer
```latex
\renewbibmacro*{date+extradate}{%
  \printlabeldateextra
}
```
Of course you'd have to find a way to get the `nodate` in square brackets if that is required, but I guess that can be done.

Reply:
Ok, I see. Thanks to this MWE, I realized that we cannot use `labeldate` at all, because that's actually against the standard...

In our edge case (an entry with no date), we should test a _nodate_ situation, because if we don't, the `urldate` as the last resort before `nodate` is printed and that's not correct. And it implies that we should override also some citing macros in `iso-authoryear.cbx` to handle this, because at the moment, the year from `urldate` is printed at the place of year in the author-year scheme: `(Appleby, 2019)` should be `(Appleby, [n.d.])`.

Reply:
You can customise from which fields `labeldate` is drawn. The default setting is
```latex
\DeclareLabeldate{%
  \field{date}
  \field{year}
  \field{eventdate}
  \field{origdate}
  \field{urldate}
  \literal{nodate}
}
```
So if you wanted to stop `urldate` from being considered, you could say
```latex
\DeclareLabeldate{%
  \field{date}
  \field{year}
  \field{eventdate}
  \field{origdate}
  \literal{nodate}
}
```

Reply:
Yes, you're right, thank you for pointing it out. I did finally the desired result in `iso-authoryear.bbx` and some tweaks in `iso.bbx` for `nodate` case, I am not so happy about that code, but didn't found any other workaround bb3e7b4.